### Big image classification test

class가 자전거인 경우에는 자전거가 포함된 이미지를 1/3 비율로 가져오고, 랜덤하게 선택한 다른 클래스의 이미지를 2/3 비율로 가져와서 테스트를 진행해보았습니다. 이 과정에서 자전거가 아닌 이미지는 클래스 0으로 예측하고, 자전거가 포함된 이미지는 클래스1로 예측하게 됩니다

In [ ]:
# 테스트 데이터 생성기
test_generator = train_data.flow_from_directory(directory=image_path,
                                               target_size=(224, 224),
                                               color_mode='rgb',
                                               batch_size=10,
                                               class_mode='categorical',
                                               shuffle=True)

Found 11671 images belonging to 12 classes.


In [ ]:
test_generator.class_indices

{'Bicycle': 0,
 'Bridge': 1,
 'Bus': 2,
 'Car': 3,
 'Chimney': 4,
 'Crosswalk': 5,
 'Hydrant': 6,
 'Motorcycle': 7,
 'Other': 8,
 'Palm': 9,
 'Stair': 10,
 'Traffic Light': 11}

In [ ]:
def create_test_generator(image_path, class_name, exclude_classes):
  # 각 라벨별 테스트 생성기 함수
  # 라벨 포함 생성기
  test_class_generator = train_data.flow_from_directory(
    directory=image_path,
    target_size=(224, 224),
    color_mode='rgb',
    batch_size=10,
    classes=[class_name],
    class_mode='categorical',
    shuffle=True
  )

  # 라벨 제외 생성기
  test_non_class_generator = train_data.flow_from_directory(
      directory=image_path,
      target_size=(224, 224),
      color_mode='rgb',
      batch_size=10,
      classes=[class_name for class_name in test_generator.class_indices.keys() if class_name not in exclude_classes],
      class_mode='categorical',
      shuffle=True
  )

  return test_class_generator, test_non_class_generator

In [ ]:
test_bicycle_generator, test_non_bicycle_generator = create_test_generator(image_path, 'Bicycle', 'Bicycle')
test_bridge_generator, test_non_bridge_generator = create_test_generator(image_path, 'Bridge', 'Bridge')
test_bus_generator, test_non_bus_generator = create_test_generator(image_path, 'Bus', 'Bus')
test_car_generator, test_non_car_generator = create_test_generator(image_path, 'Car', 'Car')
test_chimney_generator, test_non_chimney_generator = create_test_generator(image_path, 'Chimney', 'Chimney')
test_crosswalk_generator, test_non_crosswalk_generator = create_test_generator(image_path, 'Crosswalk', 'Crosswalk')
test_hydrant_generator, test_non_hydrant_generator = create_test_generator(image_path, 'Hydrant', 'Hydrant')
test_motorcycle_generator, test_non_motorcycle_generator = create_test_generator(image_path, 'Motorcycle', 'Motorcycle')
test_palm_generator, test_non_palm_generator = create_test_generator(image_path, 'Palm', 'Palm')
test_stair_generator, test_non_stair_generator = create_test_generator(image_path, 'Stair', 'Stair')
test_traffic_light_generator, test_non_traffic_light_generator = create_test_generator(image_path, 'Traffic Light', 'Traffic Light')

Found 780 images belonging to 1 classes.
Found 10891 images belonging to 11 classes.
Found 533 images belonging to 1 classes.
Found 11138 images belonging to 11 classes.
Found 1209 images belonging to 1 classes.
Found 10462 images belonging to 11 classes.
Found 3496 images belonging to 1 classes.
Found 8175 images belonging to 11 classes.
Found 124 images belonging to 1 classes.
Found 11547 images belonging to 11 classes.
Found 1240 images belonging to 1 classes.
Found 10431 images belonging to 11 classes.
Found 955 images belonging to 1 classes.
Found 10716 images belonging to 11 classes.
Found 81 images belonging to 1 classes.
Found 11590 images belonging to 11 classes.
Found 911 images belonging to 1 classes.
Found 10760 images belonging to 11 classes.
Found 211 images belonging to 1 classes.
Found 11460 images belonging to 11 classes.
Found 791 images belonging to 1 classes.
Found 10880 images belonging to 11 classes.


Found 780 images belonging to 1 classes.
Found 10891 images belonging to 11 classes.
Found 533 images belonging to 1 classes.
Found 11138 images belonging to 11 classes.
Found 1209 images belonging to 1 classes.
Found 10462 images belonging to 11 classes.
Found 3496 images belonging to 1 classes.
Found 8175 images belonging to 11 classes.
Found 124 images belonging to 1 classes.
Found 11547 images belonging to 11 classes.
Found 1240 images belonging to 1 classes.
Found 10431 images belonging to 11 classes.
Found 955 images belonging to 1 classes.
Found 10716 images belonging to 11 classes.
Found 81 images belonging to 1 classes.
Found 11590 images belonging to 11 classes.
Found 911 images belonging to 1 classes.
Found 10760 images belonging to 11 classes.
Found 211 images belonging to 1 classes.
Found 11460 images belonging to 11 classes.
Found 791 images belonging to 1 classes.
Found 10880 images belonging to 11 classes.

In [ ]:
def load_images_from_generator(generator, num_images):
    images = []
    labels = []
    num_loaded = 0

    while num_loaded < num_images:
        batch_images, batch_labels = next(generator)
        images.append(batch_images)
        labels.append(batch_labels)
        num_loaded += len(batch_images)

    images = np.concatenate(images)[:num_images]
    labels = np.concatenate(labels)[:num_images]

    return images, labels

In [ ]:
bicycle_images, bicycle_labels = load_images_from_generator(test_bicycle_generator, 200)
non_bicycle_images, non_bicycle_labels = load_images_from_generator(test_non_bicycle_generator, 400)

In [ ]:
bridge_images, bridge_labels = load_images_from_generator(test_bridge_generator, 200)
non_bridge_images, non_bridge_labels = load_images_from_generator(test_non_bridge_generator, 400)

In [ ]:
bus_images, bus_labels = load_images_from_generator(test_bus_generator, 200)
non_bus_images, non_bus_labels = load_images_from_generator(test_non_bus_generator, 400)

In [ ]:
car_images, car_labels = load_images_from_generator(test_car_generator, 200)
non_car_images, non_car_labels = load_images_from_generator(test_non_car_generator, 400)

In [ ]:
chimney_images, chimney_labels = load_images_from_generator(test_chimney_generator, 200)
non_chimney_images, non_chimney_labels = load_images_from_generator(test_non_chimney_generator, 400)

In [ ]:
cro_images, cro_labels = load_images_from_generator(test_crosswalk_generator, 200)
non_cro_images, non_cro_labels = load_images_from_generator(test_non_crosswalk_generator, 400)

In [ ]:
hydrant_images, hydrant_labels = load_images_from_generator(test_hydrant_generator, 200)
non_hydrant_images, non_hydrant_labels = load_images_from_generator(test_non_hydrant_generator, 400)

In [ ]:
moto_images, moto_labels = load_images_from_generator(test_motorcycle_generator, 81)
non_moto_images, non_moto_labels = load_images_from_generator(test_non_motorcycle_generator, 160)

In [ ]:
palm_images, palm_labels = load_images_from_generator(test_palm_generator, 200)
non_palm_images, non_palm_labels = load_images_from_generator(test_non_palm_generator, 400)

In [ ]:
stair_images, stair_labels = load_images_from_generator(test_stair_generator, 200)
non_stair_images, non_stair_labels = load_images_from_generator(test_non_stair_generator, 400)

In [ ]:
tf_images, tf_labels = load_images_from_generator(test_traffic_light_generator, 200)
non_tf_images, non_tf_labels = load_images_from_generator(test_non_traffic_light_generator, 400)

In [ ]:
class_indices = test_generator.class_indices

predictions = []
true_labels = []

class_name = 'Bicycle'

bicycle_preds = vit_model.predict(bicycle_images)

for pred, label in zip(bicycle_preds, np.argmax(bicycle_labels, axis=1)):
    print('label', label)
    if label == 0:
        predicted_class = np.argmax(pred)
        print('pred', predicted_class)
        if predicted_class == 0:
            predictions.append(1)
            true_labels.append(1)
            #print('pred : 1, true : 1')
        else:
            predictions.append(0)
            true_labels.append(1)
            #print('pred : 0, true : 1')


non_bicycle_preds = vit_model.predict(non_bicycle_images)

for pred, label in zip(non_bicycle_preds, np.argmax(non_bicycle_labels, axis=1)):
    print(label)
    if label != 0:
        predicted_class = np.argmax(pred)
        print('pred', predicted_class)
        if predicted_class == 0:
            predictions.append(1)
            true_labels.append(0)
            #print('pred : 1, true : 0')
        else:
            predictions.append(0)
            true_labels.append(0)
            #print('pred : 0, true : 0')

report_bicycle = classification_report(true_labels, predictions)

In [ ]:
print(report_bicycle)

              precision    recall  f1-score   support

           0       0.99      1.00      1.00       380
           1       0.99      0.99      0.99       200

    accuracy                           0.99       580
   macro avg       0.99      0.99      0.99       580
weighted avg       0.99      0.99      0.99       580



In [ ]:
class_indices = test_generator.class_indices

predictions = []
true_labels = []

class_name = 'Bridge'

bridge_preds = vit_model.predict(bridge_images)

for pred, label in zip(bridge_preds, np.argmax(bridge_labels, axis=1)):
    print('label', label)
    if label == 0:
        predicted_class = np.argmax(pred)
        print('pred', predicted_class)
        if predicted_class == 1:
            predictions.append(1)
            true_labels.append(1)
            #print('pred : 1, true : 1')
        else:
            predictions.append(0)
            true_labels.append(1)
            #print('pred : 0, true : 1')

non_bridge_preds = vit_model.predict(non_bridge_images)

for pred, label in zip(non_bridge_preds, np.argmax(non_bridge_labels, axis=1)):
    print(label)
    if label != 0:
        predicted_class = np.argmax(pred)
        if predicted_class == 1:
            predictions.append(1)
            true_labels.append(0)
            #print('pred : 1, true : 0')
        else:
            predictions.append(0)
            true_labels.append(0)
            #print('pred : 0, true : 0')


report_bridge = classification_report(true_labels, predictions)

In [ ]:
print(report_bridge)

              precision    recall  f1-score   support

           0       0.96      1.00      0.98       368
           1       1.00      0.93      0.96       200

    accuracy                           0.97       568
   macro avg       0.98      0.96      0.97       568
weighted avg       0.97      0.97      0.97       568



In [ ]:
from sklearn.metrics import classification_report

class_indices = test_generator.class_indices

predictions = []
true_labels = []

class_name = 'Bus'  # Change 'Bridge' to 'Bus'

bus_preds = vit_model.predict(bus_images)

for pred, label in zip(bus_preds, np.argmax(bus_labels, axis=1)):
    print(label)
    if label == 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(1)
            print('pred : 1, true : 1')
        else:
            predictions.append(0)
            true_labels.append(1)
            print('pred : 0, true : 1')

non_bus_preds = vit_model.predict(non_bus_images)

for pred, label in zip(non_bus_preds, np.argmax(non_bus_labels, axis=1)):
    print(label)
    if label != 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(0)
            print('pred : 1, true : 0')
        else:
            predictions.append(0)
            true_labels.append(0)
            print('pred : 0, true : 0')

report_bus = classification_report(true_labels, predictions)

In [ ]:
print(report_bus)

              precision    recall  f1-score   support

           0       0.99      1.00      0.99       374
           1       1.00      0.98      0.99       200

    accuracy                           0.99       574
   macro avg       0.99      0.99      0.99       574
weighted avg       0.99      0.99      0.99       574



In [ ]:
class_indices = test_generator.class_indices

predictions = []
true_labels = []

class_name = 'Car'  # Change 'Bridge' to 'Bus'

car_preds = vit_model.predict(car_images)

for pred, label in zip(car_preds, np.argmax(car_labels, axis=1)):
    print(label)
    if label == 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(1)
            print('pred : 1, true : 1')
        else:
            predictions.append(0)
            true_labels.append(1)
            print('pred : 0, true : 1')

non_car_preds = vit_model.predict(non_car_images)

for pred, label in zip(non_car_preds, np.argmax(non_car_labels, axis=1)):
    print(label)
    if label != 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(0)
            print('pred : 1, true : 0')
        else:
            predictions.append(0)
            true_labels.append(0)
            print('pred : 0, true : 0')

report_car = classification_report(true_labels, predictions)

In [ ]:
print(report_car)

              precision    recall  f1-score   support

           0       0.99      0.93      0.96       367
           1       0.88      0.97      0.93       200

    accuracy                           0.95       567
   macro avg       0.93      0.95      0.94       567
weighted avg       0.95      0.95      0.95       567



In [ ]:
class_indices = test_generator.class_indices

predictions = []
true_labels = []

class_name = 'Chimney'  # Change 'Bridge' to 'Bus'

chimney_preds = vit_model.predict(chimney_images)

for pred, label in zip(chimney_preds, np.argmax(chimney_labels, axis=1)):
    print(label)
    if label == 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(1)
            print('pred : 1, true : 1')
        else:
            predictions.append(0)
            true_labels.append(1)
            print('pred : 0, true : 1')

non_chimney_preds = vit_model.predict(non_chimney_images)

for pred, label in zip(non_chimney_preds, np.argmax(non_chimney_labels, axis=1)):
    print(label)
    if label != 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(0)
            print('pred : 1, true : 0')
        else:
            predictions.append(0)
            true_labels.append(0)
            print('pred : 0, true : 0')

report_chimney = classification_report(true_labels, predictions)

In [ ]:
print(report_chimney)

              precision    recall  f1-score   support

           0       0.95      1.00      0.98       369
           1       1.00      0.91      0.95       200

    accuracy                           0.97       569
   macro avg       0.98      0.96      0.96       569
weighted avg       0.97      0.97      0.97       569



In [ ]:
class_indices = test_generator.class_indices

predictions = []
true_labels = []

class_name = 'Crosswalk'  # Change 'Bridge' to 'Bus'

cro_preds = vit_model.predict(cro_images)

for pred, label in zip(cro_preds, np.argmax(cro_labels, axis=1)):
    print(label)
    if label == 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(1)
            print('pred : 1, true : 1')
        else:
            predictions.append(0)
            true_labels.append(1)
            print('pred : 0, true : 1')

non_cro_preds = vit_model.predict(non_cro_images)

for pred, label in zip(non_cro_preds, np.argmax(non_cro_labels, axis=1)):
    print(label)
    if label != 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(0)
            print('pred : 1, true : 0')
        else:
            predictions.append(0)
            true_labels.append(0)
            print('pred : 0, true : 0')

report_crosswalk = classification_report(true_labels, predictions)

In [ ]:
print(report_crosswalk)

              precision    recall  f1-score   support

           0       0.93      1.00      0.97       371
           1       1.00      0.87      0.93       200

    accuracy                           0.95       571
   macro avg       0.97      0.94      0.95       571
weighted avg       0.96      0.95      0.95       571



In [ ]:
class_indices = test_generator.class_indices

predictions = []
true_labels = []

class_name = 'Hydrant'  # Change 'Bridge' to 'Bus'

hydrant_preds = vit_model.predict(hydrant_images)

for pred, label in zip(hydrant_preds, np.argmax(hydrant_labels, axis=1)):
    print(label)
    if label == 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(1)
            print('pred : 1, true : 1')
        else:
            predictions.append(0)
            true_labels.append(1)
            print('pred : 0, true : 1')

non_hydrant_preds = vit_model.predict(non_hydrant_images)

for pred, label in zip(non_hydrant_preds, np.argmax(non_hydrant_labels, axis=1)):
    print(label)
    if label != 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(0)
            print('pred : 1, true : 0')
        else:
            predictions.append(0)
            true_labels.append(0)
            print('pred : 0, true : 0')

report_hydrant = classification_report(true_labels, predictions)

In [ ]:
print(report_hydrant)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00       368
           1       1.00      1.00      1.00       200

    accuracy                           1.00       568
   macro avg       1.00      1.00      1.00       568
weighted avg       1.00      1.00      1.00       568



In [ ]:
class_indices = test_generator.class_indices

class_name = 'Motorcycle'
moto_preds = vit_model.predict(moto_images)

predictions = []
true_labels = []

for pred, label in zip(moto_preds, np.argmax(moto_labels, axis=1)):
    print(label)
    if label == 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(1)
            print('pred : 1, true : 1')
        else:
            predictions.append(0)
            true_labels.append(1)
            print('pred : 0, true : 1')

non_moto_preds = vit_model.predict(non_moto_images)

for pred, label in zip(non_moto_preds, np.argmax(non_moto_labels, axis=1)):
    print(label)
    if label != 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(0)
            print('pred : 1, true : 0')
        else:
            predictions.append(0)
            true_labels.append(0)
            print('pred : 0, true : 0')

report_moto = classification_report(true_labels, predictions)

In [ ]:
print(report_moto)

              precision    recall  f1-score   support

           0       0.99      0.99      0.99       294
           1       0.97      0.95      0.96        81

    accuracy                           0.98       375
   macro avg       0.98      0.97      0.98       375
weighted avg       0.98      0.98      0.98       375



In [ ]:
class_indices = test_generator.class_indices

predictions = []
true_labels = []

class_name = 'Palm'  # Change 'Bridge' to 'Bus'

palm_preds = vit_model.predict(palm_images)

for pred, label in zip(palm_preds, np.argmax(palm_labels, axis=1)):
    print(label)
    if label == 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(1)
            print('pred : 1, true : 1')
        else:
            predictions.append(0)
            true_labels.append(1)
            print('pred : 0, true : 1')


non_palm_preds = vit_model.predict(non_palm_images)

for pred, label in zip(non_palm_preds, np.argmax(non_palm_labels, axis=1)):
    print(label)
    if label != 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(0)
            print('pred : 1, true : 0')
        else:
            predictions.append(0)
            true_labels.append(0)
            print('pred : 0, true : 0')

report_palm = classification_report(true_labels, predictions)

In [ ]:
print(report_palm)

              precision    recall  f1-score   support

           0       0.98      0.99      0.99       371
           1       0.99      0.97      0.98       200

    accuracy                           0.99       571
   macro avg       0.99      0.98      0.98       571
weighted avg       0.99      0.99      0.99       571



In [ ]:
class_indices = test_generator.class_indices

predictions = []
true_labels = []

class_name = 'Stair'  # Change 'Bridge' to 'Bus'

stair_preds = vit_model.predict(stair_images)

for pred, label in zip(stair_preds, np.argmax(stair_labels, axis=1)):
    print(label)
    if label == 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(1)
            print('pred : 1, true : 1')
        else:
            predictions.append(0)
            true_labels.append(1)
            print('pred : 0, true : 1')


non_stair_preds = vit_model.predict(non_stair_images)

for pred, label in zip(non_stair_preds, np.argmax(non_stair_labels, axis=1)):
    print(label)
    if label != 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(0)
            print('pred : 1, true : 0')
        else:
            predictions.append(0)
            true_labels.append(0)
            print('pred : 0, true : 0')

report_stair = classification_report(true_labels, predictions)

In [ ]:
print(report_stair)

              precision    recall  f1-score   support

           0       0.94      1.00      0.97       373
           1       1.00      0.89      0.94       200

    accuracy                           0.96       573
   macro avg       0.97      0.94      0.95       573
weighted avg       0.96      0.96      0.96       573



In [ ]:
class_indices = test_generator.class_indices

predictions = []
true_labels = []

class_name = 'Traffic Light'  # Change 'Bridge' to 'Bus'

tf_preds = vit_model.predict(tf_images)

for pred, label in zip(tf_preds, np.argmax(tf_labels, axis=1)):
    print(label)
    if label == 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(1)
            print('pred : 1, true : 1')
        else:
            predictions.append(0)
            true_labels.append(1)
            print('pred : 0, true : 1')


non_tf_preds = vit_model.predict(non_tf_images)

for pred, label in zip(non_tf_preds, np.argmax(non_tf_labels, axis=1)):
    print(label)
    if label != 0:
        predicted_class = np.argmax(pred)
        if predicted_class == class_indices[class_name]:
            predictions.append(1)
            true_labels.append(0)
            print('pred : 1, true : 0')
        else:
            predictions.append(0)
            true_labels.append(0)
            print('pred : 0, true : 0')

report_tf = classification_report(true_labels, predictions)

In [ ]:
print(report_tf)

              precision    recall  f1-score   support

           0       0.91      1.00      0.95       376
           1       1.00      0.81      0.90       200

    accuracy                           0.94       576
   macro avg       0.96      0.91      0.93       576
weighted avg       0.94      0.94      0.93       576



| Class          | Accuracy | F1-score about each class | F1-score about except class | count |
|----------------|----------|------------------|------------------|---------|
| Bicycle        |   0.99   |      1.00        |      0.99        | 607 |
| Bridge         |   0.96   |      0.97        |      0.94        | 574 |
| Bus            |   0.99   |      0.99        |      0.98        | 5700 |
| Car            |   0.95   |      0.96        |      0.93        | 5627
| Chimney        |   0.97   |      0.98        |      0.95        |5635|
| Crosswalk      |   0.95   |      0.96        |      0.92        |5696|
| Hydrant        |   1.00   |      1.00        |      1.00        |5693|
| Motorcycle     |   0.87   |      0.92        |      0.58        |
| Palm           |   0.88   |      0.91        |      0.81        |
| Stair          |   0.78   |      0.87        |      0.29        |
| Traffic Light  |   0.92   |      0.95        |      0.81        |
| Total          |   0.87   |      0.91        |      0.68        |



